File: frontend/web/src/utils/multi-app-cast-options.js
import { createDefaultMirrorSettings } from "../utils/mirror-cast-defaults.js";
import { buildCastPayloadFromMirrorSettings } from "../utils/build-cast-payload.js";
import { NEW_DISPLAY_CUSTOM } from "../utils/mirror-screen-constants.js";
import { resolveVdFromContent, resolveVdSize } from "./multi-app-window-layout.js";
import { suggestDpi } from "./mirror-cast-constants.js";

/**
 * Build cast options equivalent to official scrcpy:
 *   scrcpy --new-display=<W>x<H>/<dpi> --start-app=<pkg> --flex-display
 *           --no-vd-system-decorations --no-vd-destroy-content
 *
 * One device-wide scrcpy-server (web :8886); each window sends its own
 * type-101 new_display + start_app over a dedicated WebSocket.
 *
 * @param {{
 *   width?: number,
 *   height?: number,
 *   dpi?: number,
 *   packageName?: string,
 *   deviceSdk?: number,
 *   orientation?: "portrait" | "landscape",
 * }} opts
 */
export function buildMultiAppCastOptions(opts) {
  const settings = createDefaultMirrorSettings();
  const orientation = opts.orientation === "landscape" ? "landscape" : "portrait";
  const fallback = resolveVdSize(orientation);
  const width = Math.max(240, Math.round(Number(opts.width) || fallback.width));
  const height = Math.max(320, Math.round(Number(opts.height) || fallback.height));
  const dpi = Math.max(
    120,
    Math.min(640, Math.round(Number(opts.dpi) || suggestDpi(width, height))),
  );
  const packageName = String(opts.packageName || "").trim();

  // --new-display=WxH/dpi
  settings.screen.useNewDisplay = true;
  settings.screen.newDisplaySelect = NEW_DISPLAY_CUSTOM;
  settings.screen.newDisplayWidth = width;
  settings.screen.newDisplayHeight = height;
  settings.screen.newDisplayDpi = dpi;
  settings.screen.newDisplayDpiManual = true;
  // --start-app=<pkg> (launched when VD id is known)
  settings.screen.newDisplayApp = packageName;
  // --flex-display / --no-vd-system-decorations / --no-vd-destroy-content
  settings.screen.flexDisplay = true;
  settings.screen.noVdSystemDecorations = true;
  settings.screen.noVdDestroyContent = true;
  // Official server defaults (not mirror UI 5Mbps)
  settings.video.maxFps = 60;
  settings.video.bitRateMbps = 8;
  settings.video.iFrameInterval = 10;
  settings.video.resolution = "1080p";
  settings.audio.disabled = true;

  return buildCastPayloadFromMirrorSettings(settings, Number(opts.deviceSdk) || 0);
}

/**
 * @param {number} contentW
 * @param {number} contentH
 */
export function vdOptionsFromContent(contentW, contentH) {
  return resolveVdFromContent(contentW, contentH);
}

File: frontend/android/app/src/main/java/com/yiyi/cloud_phone/workspace/MultiAppCastOptions.java
package com.yiyi.cloud_phone.workspace;

import com.yiyi.cloud_phone.cast.CastPayloadBuilder;

import org.json.JSONObject;

public final class MultiAppCastOptions {
    private MultiAppCastOptions() {
    }

    public static JSONObject buildSettings(
            String packageName,
            int vdWidth,
            int vdHeight,
            int vdDpi,
            int deviceSdk
    ) throws Exception {
        JSONObject settings = CastSettingsDefaults.mirror();
        JSONObject video = settings.getJSONObject("video");
        video.put("maxFps", 60);
        video.put("bitRateMbps", 8);
        video.put("iFrameInterval", 10);
        video.put("resolution", "1080p");

        JSONObject audio = settings.getJSONObject("audio");
        audio.put("disabled", true);

        JSONObject screen = settings.getJSONObject("screen");
        CastMirrorScreenUtils.applyNewDisplaySelect(screen, CastMirrorScreenUtils.NEW_DISPLAY_CUSTOM);
        CastJson.putInt(screen, "newDisplayWidth", vdWidth);
        CastJson.putInt(screen, "newDisplayHeight", vdHeight);
        CastJson.putInt(screen, "newDisplayDpi", vdDpi);
        CastJson.putBool(screen, "newDisplayDpiManual", true);
        CastJson.putText(screen, "newDisplayApp", packageName);
        CastJson.putBool(screen, "flexDisplay", true);
        CastJson.putBool(screen, "noVdDestroyContent", true);
        CastJson.putBool(screen, "noVdSystemDecorations", true);

        settings.put("deviceSdk", deviceSdk);
        return settings;
    }

    public static JSONObject buildPayload(
            String packageName,
            int vdWidth,
            int vdHeight,
            int vdDpi,
            int deviceSdk
    ) throws Exception {
        return CastPayloadBuilder.fromMirror(
                buildSettings(packageName, vdWidth, vdHeight, vdDpi, deviceSdk),
                deviceSdk
        );
    }

    public static byte[] buildStreamParams(
            String packageName,
            int vdWidth,
            int vdHeight,
            int vdDpi,
            int deviceSdk
    ) throws Exception {
        JSONObject payload = buildPayload(packageName, vdWidth, vdHeight, vdDpi, deviceSdk);
        return CastPayloadBuilder.streamParamsFromPayload(payload);
    }
}
